### PR TITLE
Further tweaks towards full native-complex

### PR DESCRIPTION
**Pull request by mohawk2**

### CODE REVIEW

**Comment by wlmb** [comment]
Wow! Thanks! Will take a look asap.

**Comment by mohawk2** [comment]
Ha ha, you're welcome! PDL::FFTW3 and PDL::LinearAlgebra now also work with native complex numbers, so we're getting there.

**Comment by wlmb** [comment]
I'm looking at your suggestions. You did quite some work! It seems to passes all the tests. 

I wonder about a few points. You use 'all' from List::Util, but it may collide with 'all' from PDL. I guess that is one of the the reasons for using PDL::Lite instead of PDL, but maybe it will be safer (thinking about the future) not to import 'all' but use its fully qualified name (List::Util::all). 

I see you removed the annoying need of moving out of the way the real-imaginary index, as it is unneeded for the the native complex pdl's. However, the any_complex in Photonic::Util still allows PDL::Complex types, which do require the index gymnastics. The question is, did you completely eliminate PDL::Complex? Maybe any_complex should be changed to disallow PDL::Complex.

Along similar lines, maybe PDL::Complex'es and real (2,n,m...) pdl's could be coerced into native complexes.

I see that the 'i' from PDL::Complex and the r2C routines are still used in some places, and PDL::Complex'es do not play nicely yet with native complex'es. Maybe it is just in the SH modules (haven't checked carefully yet) as I see you removed the 'use PDL::Complex' from many other places.

I am still ignorant about github procedures. How can I suggest/add changes on top of your changes to discuss before committing? 

Finally, another question. the least I owe you is an acknowledgment. Is there a standard place in a git repository and in cpan to do that?

**Comment by mohawk2** [comment]
I did consider the point of `all`. If one doesn't import it, then you don't get the nice syntax and have to write it like: 

```perl
List::Util::all(sub { $_->thing }, @list)
```

The benefit of having thorough tests is that one immediately knows when there's a problem like this.

> I see you removed the annoying need of moving out of the way the real-imaginary index, as it is unneeded for the the native complex pdl's

All I did was change the order of things when one is starting from real data, by calling `r2C` later. The gymnastics are still needed for now, due to the inherent nature of `PDL::Complex`'s quite thin layer over normal PDLs (as noted in my reply to your message on the mailing list :-).

I have not completely eliminated `PDL::Complex` from this codebase yet, but it is my intent to do so. Because of those gymnastics, there are quite a number of steps required to do that in a comprehensible way - doing it all as a "big bang" is basically impossible, but by adding tests, and changing things over one at a time, it becomes feasible.

The removal so far of `PDL::Complex` was simply when it wasn't used at all in that module, which removes false positives for the further steps alluded to above :-)

As for suggesting changes, if you go to the "Files" tab of this PR, you can either add single comments right next to the lines you have questions about, or make a "review" which will send a single message when you're done, but nothing till then (and it's easy to forget as the UI isn't perfect).

Re acknowledgements, please don't worry about it. My ID shows in the commit history (which takes care of the GitHub side), and I'm not concerned about credits in the docs.

**Comment by wlmb** [comment]
Thanks for all the explanations.

A  couple of simple questions: I had some coefficients stored as perl arrays. You changed that to pdl's. Then you changed the iterations over the coefficients $coeff[$n] to iterations over $coeff->(:,$n). Is there  an advantage of one over the other?  I expected $coeff[$n] would be faster maybe I'm wrong as the structure of arrays is simpler in pdl (I used to forget that NiceSlice applies its transformations before execution). An advantage of using pdl's instead of arrays is that maybe in the future the package might be used with additional dimensions which may be threaded over, for example, for obtaining 'epsilons' for many frequencies at once. But then, for safety , I guess it would be better to remove explicitly the resulting dimension, as in $coeff->(:,($n)) with the extra parenthesis. Is this correct?

**Comment by mohawk2** [comment]
I am about 3 hours ahead of you in learning this stuff, so please bear with me if this is unclear, incomplete, or wrong! When you use NiceSlice, it's the same as calling the `slice` method with the string, so `pdldoc slice` should shed some light. When you slice a dimension `x` that means index by `x`, but keep that dimension visible in the result. If you use `(x)`, the dimension is not visible in the result, i.e. the dimensions of that result are fewer.

My motivation for that particular change is that it's a step towards eliminating the `for` loop there, and having a vectorised calculation. If I understood LU decomposition and substitution better I'd have included that already. As it is, I'm a bit unclear on how to change the matrix arrangement there from its mapping of a linear (one-dimensional) representation of the lower triangle of that matrix into a rectangular form. It's probably done by making an intermediate PDL with 3 columns being the linear index `n`, then the x `i` and y `j`, then doing `$tensor->indexND($inter->(1:2)) = $backsub->($inter->(0))`. One can probably then just do `$tensor += $tensor->transpose`.

Basically, vectorised calculations are a great deal faster than doing things in Perl! And you'd get being able to do extra dimensions just by calculating the slice arguments as you're already doing elsewhere.

And just to be very explicit, the first dimension for `$tensor` is needed as that's the `PDL::Complex` extra dimension, which will go away once this thing changes over to native complex, which none of this PR actually executes yet.

**Comment by mohawk2** [comment]
By the way, to be sufficiently sure we don't blow up the second-harmonic code, I'm trying to figure out plausible minimal tests for it. I'm having to learn about dielectric thingies in order to figure plausible values for the densities of the two materials for a `Photonic::LE::NR2::SHP`. It would save me some effort if you could help with plausible values for `densityA` and `densityB` :-D

**Comment by wlmb** [comment]
epsilonA=1, densitaA=0 (means vacuum), epsilonB=whatever complex (with nonnegative imaginary part, and positive real part for faster/better convergence) and denstityB=1 (the results are normalized to density B if I recall correctly, so the actual (real positive) value doesn't matter).

**Comment by wlmb** [comment]
Thinking about the iteration vs. vectorization, the part of the computation that requires 'states' might have to be iterated, as the states might consume a lot of memory, so allow writing them to disk and bringing them back one at a time in sequence when needed. Thus, the calculation of fields and the reorthogonalizations have to be done iteratively. (I don't remember if there are disk based pdl's somewhere); but other calculations, such as the tensor stuff could be vectorized (I'll think about it). For example, the lentz stuff calculates a continued fraction which is one element of the solution of a tridiagonal system. An alternative would be to obtain the complete solution, which could be vectorized, and then single out the desired element. It may be faster. Maybe it would be nice to offer both choices throught some configuration field and let the users choose.

**Comment by mohawk2** [comment]
Thanks! The `P2LMCalt` is blowing up because it uses `states` which the role blows up on. I'll have to convert it to use an iterator, and delete `linearCombine`.

You're quite right about the tradeoffs of space and time, but the situation we were discussing above was about things that are all in memory, and need calculating, so that might as well be done as efficiently as possible :-) I can't comment on the Lentz stuff, but you may well be right. Does LAPACK offer any functions to do those calculations? (See PDL::LinearAlgebra as well to see if they're already available for no extra effort :-)

**Comment by wlmb** [comment]
P2LMCalt was not been updated when I introduced the iterators because it never did work (and I don't know why). Maybe it would be best to just comment it out. 

The call to lentzCF could be replaced by a call to cgtsv and throwing away the result except for its first element. Some information would be lost though, as currently the calculation may converge and terminate before all the elements of the tridiagonal matrix are used, which is a sign of convergence. 

I haven't found a package offering the Lentz algorithm; I adapted my routine from the Numerical Recipes books which have C and Fortran versions, which could be interfaced to PDL... As this is a relatively inexpensive contribution to the calculations I was satisfied with a PERL version. So maybe it would be reasonable to keep the iterations using pdl's instead of arrays and indexing with NiceSlice. This would be compatible with allowing additional indices in the future, I guess. 

By the way, how inefficient are the routines defined through 'thread_define'?

**Comment by mohawk2** [comment]
`thread_define` - I'm afraid I don't know! I've seen a reference to it in the code / docs, but not had a go yet. The usual answer to that sort of question, as here is - construct a plausible example that runs for a decent amount of time (say 10 seconds), and try various different techniques to learn which is fastest.

**Comment by mohawk2** [comment]
My usual approach is not to delete code or to throw away information without a very good reason. I'm having a go at fixing `P2LMCalt` to at least run.

**Comment by mohawk2** [comment]
Got the P2LMCalt working. I don't know if it's mathematically correct! Now I'm adding other little tests (and occasional fixes) towards getting 100% test coverage.

Could you let me know if there are any problems with this pull request that mean you'd like changes? It's getting a bit big and I'd like to make the actual native-complex changes in a separate PR, so if there aren't problems you might merge this?

**Comment by wlmb** [comment]
I don't see any problems. Will merge. It still makes me nervous the replacement of $array[$n] by $pdl->(:,$n) without the parenthesis, as in $pdl ->(:,($n)). It seems safe for now (and it passes the tests), but if additional threadable dimensions are allowed in the future, it will have to change. For example, when computing fields (see line 217 of LE/NR2/Field.pm) 
        my $EnGPsi_G=$GPsi_G*$Es->(:,$n); #En ^G|psi_n>
	$field_G+=$EnGPsi_G;
$GPsi_G currently has dimensions 2 (real or imag), ndims (the components of the vector field), nx, ny... (the field arguments) while $Es->(:,$n) has dimensions 2(real or imag) and 1 (a dummy arg). If I were to add threadable additional dimensions t1, t2..., then $GPsi_G would be 2,ndims,nx,ny...t1,t2... and $Es->(:,$n) would be 2,1,t1,t2... and the product would fail. Thus, at some moment we should change $Es->(:,$n) by $Es->(:,($n),*1,*1...) where there are as many *1's as there are dimensions in space. No urgency though.

**Comment by mohawk2** [comment]
Thanks for merge! It seems to me that there are no risk-free options, but with 100% tests any problems will be instantly apparent. As mentioned above, the "(0)" means that it's equivalent to an array-lookup, which is why no other changes were needed in that bit of the code. If more threadable dimensions were added, it would still be equivalent. Slicing is the best bit about PDL, please try the command-line `pdl` (or `pdl2`) and try things like:

```
pdl> p $x = sequence 2, 3

[
 [0 1]
 [2 3]
 [4 5]
]

pdl> p $x->slice(',(1)')
[2 3]
pdl> p $x->slice(',*')

[
 [
  [0 1]
 ]
 [
  [2 3]
 ]
 [
  [4 5]
 ]
]
```

**Comment by wlmb** [comment]
_If more threadable dimensions were added, it would still be equivalent._ 

Not if the states from which the field is built and the coefficients for the linear combination both have extra threadable dimensions, even if they are not the same, as the field has additional non-extra dimensions.

For example: multiply one of several scalar times a field.
```
pdl> $field=sequence(3,2,2,2)  # real field in 3d with 2 values for x, y and z    
pdl> $coefficients=sequence(4) # array of scalar coefficients
pdl> $combine=$coefficients(1)*$field # term of the linear  combination as of now
pdl> p $combine->info
PDL: Double D [3,2,2,2] 
```
Now the same but with a threadable extra dimension:
```
pdl> $field_threaded=sequence(3,2,2,2,5) # add tjread dimension of size 5 to the field
pdl> $coefficients_threaded=sequence(4,5) # and to the coefficients
pdl> $combine_threaded=$coefficients_threaded(1)*$field_threaded # this is the error I mentioned 
Runtime error: PDL: PDL::Ops::mult(a,b,c): Parameter 'b':
  Mismatched implicit thread dimension 1: size 5 vs. 2
There are 3 PDLs in the expression; 5 thread dims.
   PDL IN EXPR.    THREAD DIMS
   #  0 (normal):        1       5
   #  1 (normal):        3       2       2       2       5
   #  2 (null)
... at (eval 405) line 6.
pdl> $combine_threaded=$coefficients_threaded((2),*1,*1,*1,*1)*$field_threaded # solved with () and dummies, one for the cartesian index and one for each spatial dimension 
pdl> p $combine_threaded->info  
PDL: Double D [3,2,2,2,5] 
```
So, to safely multiply a scalar with a vector field I have to explicitly add (ndims+1) dummy dimensions to the scalar. Then I can safely add extra dimensions if desired afterwards.
